File: backend/app/api/workflow.py
"""Workflow Studio API: view and edit the agent orchestration per domain.

The planner's pipeline topology is fixed, but WHICH specialists run at each
decision point (the roster) is configuration. This endpoint exposes the graph
(stages, always-on nodes, selectable specialists, and each decision point's
roster) and lets an org pin a custom roster per decision point. The override is
stored in the same per-org ``pack_overrides`` blob the Rules editor uses (under
a ``rosters`` key) and is honored by the planner at run time via
``set_roster_overrides``. Org-scoped; one tenant never sees another's edits.
"""

from __future__ import annotations

from typing import Any, Dict, List

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from app.api._org import current_org
from app.packs.loader import load_pack, reset_pack_org, set_pack_org
from app.repositories import pack_overrides as overrides_repo

router = APIRouter(prefix="/workflow", tags=["workflow"])


class WorkflowIn(BaseModel):
    # {decision_point_key: [capability, ...]}. An empty/omitted map clears the
    # roster override and reverts to the pack's configured rosters.
    rosters: Dict[str, List[str]] = Field(default_factory=dict)


def _specialists() -> List[Dict[str, Any]]:
    """The selectable specialist roster the planner draws from, in run order."""
    from app.graph.planner import _ALWAYS_ON, _PIPELINE, _SEQUENCE
    from app.packs.registry import AGENTS

    out: List[Dict[str, Any]] = []
    for cap in _SEQUENCE:
        if cap in _PIPELINE and AGENTS.has(cap):
            card = AGENTS.find(cap)
            out.append(
                {
                    "capability": cap,
                    "description": card.description if card else "",
                    "always_on": cap in _ALWAYS_ON,
                }
            )
    return out


def _pack_for(org_id: str, domain: str):
    token = set_pack_org(org_id)
    try:
        return load_pack(domain)
    finally:
        reset_pack_org(token)


async def _view(org_id: str, domain: str) -> Dict[str, Any]:
    from app.graph.planner import _ALWAYS_ON, _SEQUENCE, _deterministic_roster

    pack = _pack_for(org_id, domain)
    blob = await overrides_repo.get_overrides(org_id, domain)
    roster_ovr = (blob or {}).get("rosters") or {}

    decision_points: List[Dict[str, Any]] = []
    for key, dp in (pack.decision_points or {}).items():
        base_roster, base_rationale = _deterministic_roster(pack, key)
        override = roster_ovr.get(key)
        decision_points.append(
            {
                "key": key,
                "label": getattr(dp, "label", key),
                "signals": list(
                    getattr(dp, "signals", None) or getattr(dp, "signal_types", None) or []
                ),
                "base_roster": base_roster,
                "rationale": base_rationale,
                "roster": list(override) if override else base_roster,
                "overridden": bool(override),
            }
        )

    return {
        "domain": domain,
        "domain_name": getattr(pack, "name", domain),
        "sequence": list(_SEQUENCE),
        "always_on": sorted(_ALWAYS_ON),
        "specialists": _specialists(),
        "decision_points": decision_points,
        "has_override": bool(roster_ovr),
    }


@router.get("/{domain}")
async def get_workflow(
    domain: str, org_id: str = Depends(current_org)
) -> Dict[str, Any]:
    """Return the orchestration graph and each decision point's effective roster."""
    return await _view(org_id, domain)


@router.put("/{domain}")
async def put_workflow(
    domain: str, body: WorkflowIn, org_id: str = Depends(current_org)
) -> Dict[str, Any]:
    """Save the org's per-decision-point roster overrides (merged onto rules)."""
    blob: Dict[str, Any] = await overrides_repo.get_overrides(org_id, domain) or {}
    cleaned = {
        key: [str(c) for c in caps]
        for key, caps in (body.rosters or {}).items()
        if caps
    }
    if cleaned:
        blob["rosters"] = cleaned
    else:
        blob.pop("rosters", None)

    if blob:
        await overrides_repo.upsert(org_id, domain, blob)
    else:
        await overrides_repo.delete(org_id, domain)
    return await _view(org_id, domain)

File: frontend/lib/api/workflow.ts
// Typed client for the Workflow Studio.
//
// The backend exposes the agent orchestration graph for a domain (the full
// specialist sequence, the always-on specialists, the selectable roster, and
// the decision points an org can tune). GET returns the effective view (base
// pack merged with the org override); PUT persists per-decision-point rosters
// as the org override. Reads degrade to a null view when the backend is
// unreachable so the page can show an error state rather than crash. Requests
// carry credentials so they work behind an org-scoped backend.

import { API_BASE, ApiError } from "@/lib/api";
import { authHeaders } from "@/lib/auth";

// A selectable specialist candidate, in run order.
export interface WorkflowSpecialist {
  capability: string;
  description: string;
  always_on: boolean;
}

// A decision point the planner routes through, with its effective roster.
export interface WorkflowDecisionPoint {
  key: string;
  label: string;
  signals: string[];
  base_roster: string[];
  rationale: string;
  roster: string[];
  overridden: boolean;
}

// The full effective workflow view for a domain.
export interface WorkflowView {
  domain: string;
  domain_name: string;
  sequence: string[];
  always_on: string[];
  specialists: WorkflowSpecialist[];
  decision_points: WorkflowDecisionPoint[];
  has_override: boolean;
}

// The PUT payload: a map of decision point key to its desired roster. Omitting
// a decision point (or sending an empty map) reverts it to the pack default.
export interface WorkflowRosters {
  [decisionPointKey: string]: string[];
}

/** Fetch the effective workflow view for a domain. */
export async function getWorkflow(
  domain: string,
  signal?: AbortSignal,
): Promise<WorkflowView> {
  const res = await fetch(
    `${API_BASE}/workflow/${encodeURIComponent(domain)}`,
    {
      headers: authHeaders({ Accept: "application/json" }),
      credentials: "include",
      cache: "no-store",
      signal,
    },
  );
  if (!res.ok) {
    throw new ApiError(`GET /workflow/${domain} failed (${res.status})`, res.status);
  }
  return (await res.json()) as WorkflowView;
}

/** Persist per-decision-point rosters as the org override and return the view. */
export async function putWorkflow(
  domain: string,
  rosters: WorkflowRosters,
): Promise<WorkflowView> {
  const res = await fetch(
    `${API_BASE}/workflow/${encodeURIComponent(domain)}`,
    {
      method: "PUT",
      headers: authHeaders({ "Content-Type": "application/json" }),
      credentials: "include",
      body: JSON.stringify({ rosters }),
    },
  );
  if (!res.ok) {
    throw new ApiError(`PUT /workflow/${domain} failed (${res.status})`, res.status);
  }
  return (await res.json()) as WorkflowView;
}
